File: State/VendingMachine.cpp
#include <iostream>
// State pattern example: Vending Machine
class IVendingMachineState
{
public:
  virtual void handleRequest() = 0;
  virtual ~IVendingMachineState() { std::cout << "delete" << std::endl; }
};

class ReadyState : public IVendingMachineState
{
public:
  void handleRequest() override
  {
    std::cout << "Ready state: Please select a product." << std::endl;
  }
};

class ProductSelectedState : public IVendingMachineState
{
public:
  void handleRequest() override
  {
    std::cout << "Product selected state: Processing payment." << std::endl;
  }
};

class PaymentPendingState : public IVendingMachineState
{
public:
  void handleRequest() override
  {
    std::cout << "Payment pending state: Dispensing product." << std::endl;
  }
};

class OutOfStockState : public IVendingMachineState
{
public:
  void handleRequest() override
  {
    std::cout << "Out of stock state: Product unavailable. Please select another product." << std::endl;
  }
};

class VendingMachineContext
{
  IVendingMachineState *state = nullptr;

public:
  VendingMachineContext() : state(nullptr) {}
  ~VendingMachineContext()
  {
    delete state;
  }
  void setState(IVendingMachineState *state_)
  {
    if (state != nullptr)
    {
      delete state;
    }
    this->state = state_;
  }

  void request()
  {
    state->handleRequest();
  }
};

int main()
{
  // Create context on the stack
  VendingMachineContext vendingMachine;

  // Set initial state
  vendingMachine.setState(new ReadyState());

  // Request state change
  vendingMachine.request();

  // Change state
  vendingMachine.setState(new ProductSelectedState());

  // Request state change
  vendingMachine.request();

  // Change state
  vendingMachine.setState(new PaymentPendingState());

  // Request state change
  vendingMachine.request();

  // Change state
  vendingMachine.setState(new OutOfStockState());

  // Request state change
  vendingMachine.request();
}

File: State/VendingMachine.java
package State;

// State interface
interface IVendingMachineState {
  void handleRequest();
}

// Concrete State: Ready
class ReadyState implements IVendingMachineState {
  @Override
  public void handleRequest() {
    System.out.println("Ready state: Please select a product.");
  }
}

// Concrete State: Product Selected
class ProductSelectedState implements IVendingMachineState {
  @Override
  public void handleRequest() {
    System.out.println("Product selected state: Processing payment.");
  }
}

// Concrete State: Payment Pending
class PaymentPendingState implements IVendingMachineState {
  @Override
  public void handleRequest() {
    System.out.println("Payment pending state: Dispensing product.");
  }
}

// Concrete State: Out of Stock
class OutOfStockState implements IVendingMachineState {
  @Override
  public void handleRequest() {
    System.out.println("Out of stock state: Product unavailable. Please select another product.");
  }
}

// Context class that holds the current state
class VendingMachineContext {
  private IVendingMachineState state;

  public VendingMachineContext() {
    this.state = null; // Start with no state
  }

  public void setState(IVendingMachineState state) {
    this.state = state;
  }

  public void request() {
    if (this.state != null) {
      this.state.handleRequest();
    }
  }
}

// Main class to demonstrate the State pattern
public class VendingMachine {
  public static void main(String[] args) {
    // Create context
    VendingMachineContext vendingMachine = new VendingMachineContext();

    // Set initial state
    vendingMachine.setState(new ReadyState());
    vendingMachine.request();

    // Change state
    vendingMachine.setState(new ProductSelectedState());
    vendingMachine.request();

    // Change state
    vendingMachine.setState(new PaymentPendingState());
    vendingMachine.request();

    // Change state
    vendingMachine.setState(new OutOfStockState());
    vendingMachine.request();
  }
}
